src/ebonite/build/runner/base.py

```diff
@@ -1,20 +1,74 @@
 from abc import abstractmethod
-from typing import Generator
+from typing import Generator, Type
+
+from ebonite.core.objects import Image, RuntimeEnvironment, RuntimeInstance
 
 
 class RunnerBase:
     @abstractmethod
-    def run(self, instance, **kwargs):
+    def instance_type(self) -> Type[RuntimeInstance.Params]:
+        """
+        :return: subtype of :class:`.RuntimeInstance.Params` supported by this runner
+        """
+        pass  # pragma: no cover
+
+    @abstractmethod
+    def create_instance(self, name: str, **kwargs) -> RuntimeInstance.Params:
+        """
+        Creates new runtime instance on given name and args
+
+        :param name: name of instance to use
+        :return: created :class:`.RuntimeInstance.Params` subclass instance
+        """
         pass  # pragma: no cover
 
     @abstractmethod
-    def is_running(self, instance) -> bool:
+    def localhost_env(self) -> RuntimeEnvironment.Params:
```

Comment:
please elaborate on why is this needed

Reply:
This method is called by `Ebonite.run_service` to create localhost `Environment` if no explicit `Environment` given. For example, our testing code has never specified any environments. This solution dramatically simplifies things in testing / REPL scenarios. For me this has more usefulness than effort to implement.

Reply:
i agree that things must be simplified, but my intuition tells me there is a better way. give me some time to think about it, if i wont come up with something, we'll do it your way